rustfmt-core/rustfmt-lib/src/expr.rs

```diff
@@ -42,6 +42,17 @@ impl Rewrite for ast::Expr {
     }
 }
 
+impl Rewrite for Option<ptr::P<ast::Expr>> {
```

Comment:
I am a bit worried about this `impl` as the meaning of `Option<ptr::P<ast::Expr>>` could vary depending on the situation. Returning an empty string against `None` makes sense in the context of `PatKind::Range`, but may not in other usages of `Option<ptr::P<ast::Expr>>` in AST. I think it is safer to make this a function/closure specific to implementing the pattern formatting, rather than making it a global `impl Rewrite`.

Reply:
Yup, makes perfect sense! I've moved it so that it's only scoped to PatKind::Range since as you mentioned this is specifically and exclusively to handle ranges that may be half open.

Reply:
Hmm, it looks like `impl Trait for Struct` always has a global scope, even if it is enclosed inside a function.

I think that the better approach here is to define a new type and implement the `Rewrite` trait on that type. This way, the scope of `impl Rewrite` is limited to that type and better conveys the intention.

E.g., something like the following:

```rust
struct RangeOperand {
    expr: Option<ptr::P<ast::Expr>>,
}

impl Rewrite for RangeOperand {
    // implement Rewrite
}

// Use RangeOperand instead of Option<ptr::P<ast::Expr>>
rewrite_pair(
    RangeOperand::new(lhs),
    RangeOperand::new(rhs),
    PairParts::infix(&infix),
    context,
    shape,
    SeparatorPlace::Front,
)
```

Reply:
Ahh you are right sorry. The `.rewrite(...)` would not have ben available outside the function scope for other AST items, but the impl def is indeed global which would have locked rewrite with a range pattern specific lens. 

Have updated it accordingly as suggested, though I elected to use a tuple struct as this felt like a newtype scenario given the single field

Reply:
SGTM, thank you for the update! Sorry, but would you resolve the conflict, please?

Reply:
Rebased and resolved, and an additional tiny commit to fix a formatting issue that snuck through before the self-tests were all back up and passing ✔️